mnml with infoline

jedahan opened the pull request:
I did another rebase, now with a GEOMETRY_INFO array which will run on an empty command, and has no known issues - I think it is stable enough for a beta.

This supersedes #233

Commit:
Change api from plugins to functions

This is a huge rewrite based on the idea that PROMPT and RPROMPT
can just be the joined output of functions.

Changes in this commit include:

- changing GEOMETRY_PROMPT_PLUGINS{,_SECONDARY} to GEOMETRY_[R]PROMPT
- adding GEOMETRY_INFO which runs when pressing enter with an empty command
- splitting the path plugin to status and path
- folding all libraries from `lib` into geometry.zsh
- prefixing all geometry commands with geometry::{wrap,prompt,rpompt,info,etc...}
- replaced git dependency with curl dependency for async.zsh
- renaming `_geometry_colorize` to `ansi`
- renaming all environment variables to be GEOMETRY_<FUNCTION>_<SYMBOL/COLOR>_THING
- removing PROMPT_GEOMETRY_SHOW_RPROMPT and PROMPT_GEOMETRY_RPROMPT_ASYNC (default on)
- removed global concept of pinning, functions can decide to export a _PIN env variable
- removed UTF-8 character width bug workaround

Improvements

- GEOMETRY_[R]PROMPT arrays are now dynamic  - functions can be added/removed/changed without reload
  - order of loading functions does not matter

Commit:
migrate hg

Breaking changes:

GEOMETRY_COLOR_HG_DIRTY => GEOMETRY_HG_COLOR_DIRTY
GEOMETRY_COLOR_HG_CLEAN => GEOMETRY_HG_COLOR_CLEAN
GEOMETRY_COLOR_HG_BRANCH => GEOMETRY_HG_COLOR_BRANCH

GEOMETRY_SYMBOL_HG_DIRTY => GEOMETRY_HG_SYMBOL_DIRTY
GEOMETRY_SYMBOL_HG_CLEAN => GEOMETRY_HG_SYMBOL_CLEAN
GEOMETRY_SYMBOL_HG_SEPARATOR => GEOMETRY_HG_SYMBOL_SEPARATOR

Commit:
migrate hostname

- removed GEOMETRY_PLUGIN_HOSTNAME_PREFIX
- removed GEOMETRY_PLUGIN_HOSTNAME_SUFFIX

- added GEOMETRY_HOSTNAME_SEPARATOR
- added GEOMETRY_HOSTNAME_HIDE_ON to hide on specified host (default localhost)

Commit:
migrate jobs

Breaking changes:

- GEOMETRY_COLOR_JOBS => GEOMETRY_JOBS_COLOR
- GEOMETRY_SYMBOL_JOBS => GEOMETRY_JOBS_SYMBOL

Commit:
migrate kube

Breaking changes:

- GEOMETRY_COLOR_KUBE => GEOMETRY_KUBE_COLOR
- GEOMETRY_SYMBOL_KUBE => GEOMETRY_KUBE_SYMBOL

Commit:
migrate rustup

- breaking change: GEOMETRY_X_RUSTUP => GEOMETRY_RUSTUP_X
- removed dependency on rustup_prompt_helper

Commit:
migrate virtualenv

- breaking changes: GEOMETRY_X_VIRTUALENV => GEOMETRY_VIRTUALENV_X

Commit:
migrate ruby

- breaking changes: GEOMETRY_X_RUBY => GEOMETRY_RUBY_X

Commit:
migrate node

- breaking changes: GEOMETRY_X_NODE => GEOMETRY_NODE_X
- added GEOMETRY_NODE_PIN for pinning

Commit:
migrate docker_machine

- breaking changes: GEOMETRY_X_DOCKER_MACHINE => GEOMETRY_DOCKER_MACHINE_X

Commit:
migrate exec_time

- breaking change: PROMPT_GEOMETRY_COMMAND_EXEC_TIME => GEOMETRY_EXEC_TIME_PATIENCE

Commit:
migrate git

- breaking change: GEOMETRY_X_GIT => GEOMETRY_GIT_X
- breaking change: default rebase symbol no longer needs powerline

Commit:
Update and lowercase documentation

This also moves the function documentation to the top level

Commit:
path: inline

Commit:
status: change root symbol to upside-down triangle

Commit:
documentation: merge functions into contributing

Commit:
table-ify installation instructions

Commit:
reset FAQ for v2.0

Commit:
link to contribution guide from readme

Commit:
simplify installation table

Commit:
remove unused gitmodules

Commit:
Fix #239 zcompile errors (thanks @psprint!)

frm commented on contributing.md:
```diff
@@ -0,0 +1,129 @@
+
+Details about creating a new function are listed [in the readme][].
+
+You can visit [the functions wiki page][] for inspiration and examples.
```
Missing links here

jedahan replied:
They should be at the bottom of the markdown file, but both are 404'ing. Fixing now.

frm commented on contributing.md:
```diff
@@ -0,0 +1,129 @@
+  # Rendering code here
+  command_i_need_to_work --version
+}
+
```
Extra newline

jedahan replied:
Fixed in c6a242a07fcf6fcc82eeb6dbc5d45c648e1ba088

frm commented on geometry.zsh:
```diff
@@ -1,56 +1,90 @@
+  async_register_callback geometry geometry::rprompt
+  async_job geometry geometry::wrap GEOMETRY_RPROMPT $PWD
+}
+add-zsh-hook precmd geometry::prompt
```
For this to work we need to ensure `add-zsh-hook` is loaded via `autoload -U add-zsh-hook`.

Also, missing newline here.

jedahan replied:
Fixed in c6a242a07fcf6fcc82eeb6dbc5d45c648e1ba088

frm commented on geometry.zsh:
```diff
@@ -1,56 +1,90 @@
+for fun in "${GEOMETRY_ROOT}"/functions/geometry_*.zsh; do . $fun; done
+
+(( $+functions[ansi] )) \
+  && { >&2 echo "ansi function already defined, cannot use geometry"; exit -1 } \
```
The `exit -1` is probably a bad idea. I do a lot of resourcing of my dotfiles and by doing this step, it closes down my shell.

jedahan replied:
should i remove the `exit -1` for loading the zsh-async library as well?

jedahan replied:
Fixed in c6a242a07fcf6fcc82eeb6dbc5d45c648e1ba088

frm commented on functions/geometry_status.zsh:
```diff
@@ -0,0 +1,26 @@
+
+  local color=GEOMETRY_STATUS_COLOR symbol=GEOMETRY_STATUS_SYMBOL
+  [[ $UID = 0 || $EUID = 0 ]] && symbol+=_ROOT && color+=_ROOT
+  (( $GEOMETRY_LAST_STATUS )) && symbol+=_ERROR
```
We should add a color for error. e.g: I use the same symbol for error as for the prompt but with different colors. I find that visual cue to be more commonly used and easier to process.

jedahan replied:
Fixed in c6a242a07fcf6fcc82eeb6dbc5d45c648e1ba088

frm reviewed (changes requested):
Sorry for being away, but I'm in the middle of a relocation process.

I left some comments but I have one final question: are these changes supposed to include a function override similar to #226 ? Given [this](https://github.com/geometry-zsh/geometry/pull/226#issuecomment-427457373) comment. Or should that be added later on?

Either way, huge work. Updating my zshrc was very straightforward with these changes! 👏

Commit:
Incorporate review feedback

jedahan commented:
Regarding #226, I'd like to see that as a later PR, perhaps v2.1. I need to re-read it to understand what #226 does that cannot already be achieved by this PR.

Commit:
Fix geometry_rustup

justmendes commented:
The main thing would be changing the order of what's inside a plugin. A good example would be the git plugin which is rather complex.

frm commented on functions/geometry_git.zsh:
```diff
@@ -0,0 +1,118 @@
+_conflicts() {
+  conflicts=$(git diff --name-only --diff-filter=U)
+
+  [[ -n "$conflicts" ]] && return
```
This should be `-z` otherwise it will always show the conflict indicator when there are none.

jedahan replied:
Fixed in df46c0c49b9933bc3eafad162aab82cc8fbbef1a

Commit:
fix conflicts logic

Commit:
cleanup redundant bindkey option

jedahan commented:
> The main thing would be changing the order of what's inside a plugin. A good example would be the git plugin which is rather complex.

Yeah I think it would be nice to revisit the git plugin in a future pr. I'll start a tracking issue now.

Commit:
Suppress chpwd side-effects in geometry::wrap

Normally chpwd hooks will be fired with a cd, but -q
suppresses that. This was an issue debugged by @psprint
in #239 involving load order and what hooks are fired on
start

Commit:
Remove the weird GEOMETRY_GIT_SHOW_ variables.

We now expose all the git helper functions to let people
customize what functions and what order of them are shown.

jedahan commented:
@justmendes I made it so you can change the order of the git plugin simply by calling the helper functions directly. Its not perfect but seems good.

Commit:
three more lines of code removed

Commit:
Fix #224 git dirtiness check in subdirectory (thanks @dmhacker!)

jedahan commented:
Just a heads up, this would close #226  #225  #224 #239 #220 #207 (though this might be closed regardless)

Commit:
Add npm_package_version and rust_version from @drager

@drager contributed two useful plugins that @jedahan ported from
[#228](https://github.com/geometry-zsh/geometry/issues/228)

geometry_npm_package_verion and geometry_rust_version

Commit:
quote more aggressively

Commit:
remove redundant local

Commit:
Add initial testing dockerfile

Commit:
Move add-zsh-hook autoloading to above functions

Commit:
specify output path for async.zsh

Commit:
make sure exec_time loads add-zsh-hook if needed

justmendes commented:
Sorry, been relocating for the past month. Let's close this, it's a large refactor and we should work on it off master for bug fixing.

jedahan commented:
Congrats on the move!

I think because its a big change, I would add a few more todos before merging:
- [ ] record screencast showing function-based approach
- [ ] add a `deprecated.zsh` plugin that warns the user if they are using any of the old variable names that the prompt has changed and how to fix it (or automatically guessing)
- [ ] write a release announcement for r/zsh r/commandline r/unixporn etc...

Anything else?

justmendes commented:
>  record screencast showing function-based approach

This is the function override feature, right?

>  add a deprecated.zsh plugin that warns the user if they are using any of the old variable names that the prompt has changed and how to fix it (or automatically guessing)

I like this, we should probably do an upgrading guide and point them there.

MarioRicalde commented:
Hello. I was looking to install geometry prompt today. I was wondering if this is the branch I should be working with since you guys seem like you're going to make a release soon?

Let me know if I can help with anything.

MarioRicalde commented:
When attempting to use your branch with `zplug` I get the following warning.

```bash
WARNING:  Failed to load "~/.zplug/repos/jedahan/geometry/geometry.zsh-theme" (jedahan/geometry)
```

The code in `.zshrc` is:

```bash
zplug "jedahan/geometry", at:mnml, as:theme
```

jedahan commented:
@MarioRicalde can you open a new issue that its not working with zplug?

MarioRicalde commented:
@jedahan the one in master of this repo is working without warnings. Should I open an issue over here or your repo

Also, I did some research and found this:

https://github.com/zplug/zplug/issues/424#issuecomment-320523980

The prompt seems to work, you just get that warning.

For the time being what I did was clone the repo and source it manually.

MarioRicalde commented:
Also, I'm not sure if this is an issue. 

![image](https://user-images.githubusercontent.com/125900/52727653-4ebcf500-2f7b-11e9-8017-74fe3ada5222.png)

jedahan commented:
yes please open new issues on my repo, thanks for all the reporting!

justmendes commented:
On the previous topic:

> record screencast showing function-based approach

do you need help with this? I have recording tools if needed.

> add a deprecated.zsh plugin that warns the user if they are using any of the old variable names that the prompt has changed and how to fix it (or automatically guessing)

I can write a script that warns for any defined & deprecated variables if you want

> write a release announcement for r/zsh r/commandline r/unixporn etc...

can also do this, if needed. Let me know what I can help in

MarioRicalde commented:
@jedahan you don't have issues enabled on your repo 🌵 

@justmendes let me know if there's anything I can do to help too, I'm down to help with this.

MarioRicalde commented:
Since the repo still has issues closed, and I'm finding more stuff, I'll drop it here for now:

the "no commits" functionality on the git function doesn't seem to be working as intended out of the box.

![image](https://user-images.githubusercontent.com/125900/52730352-c2153580-2f80-11e9-8c0a-e714e3e30a59.png)

After reading the code I found about that variable that once set it suddenly works, although it has some issues with formatting, `no::commits` instead of `no commits`.

Commit:
fix no-commits for geometry_git_time

MarioRicalde commented:
Sweet, I saw you added issues and I saw you just fixed the issue I just reported ;)

jedahan commented:
I enabled issues for my repo @MarioRicalde, and added the title issue

@justmendes The thing that would be most helpful is drafting an announcement. I don't know why I avoid writing it but I do 🤷‍♂️ 

@justmendes or @MarioRicalde A `geometry_deprecated()` function that would look for set variables and 1. warn users of the new names, 2. pointed to the new readme or a migration guide on a wiki would be also very helpful

I can make a short svg of the function approach

Commit:
Interpolate vars in set_title, fixes #2

MarioRicalde commented:
@jedahan @justmendes  I can help provide proofreading or fluffing of content if you give me some points you want to touch, I love writing.

I can also help with graphics since I do UI/UX stuff at my day job.

I'm new to the project, but I'll contribute with whatever is possible, just point me in the direction of what you want to be done and some guidance and I'll work on it. For the time being, I'm going to review your stuff @jedahan and blast you with issues.

Commit:
be less strict about sourcing async

justmendes commented:
@jedahan I can write and also do some custom themes, like the one I'm using now, based on an older revision of this branch.

@MarioRicalde do you want to start by doing that deprecated function? I think you might benefit of doing that as it requires you to have an overview of the codebase. It would be a good entry point into the project.

Since you mentioned graphics, we had #164 open for a while. Would you be interested in reopening this? It would be something simple and minimal, as these are the ideas of the theme.

Commit:
define geometry_hg even if hg does not exist

Commit:
geometry_exec_time: exit successfully if the file is empty or doesnt exist

Commit:
Add GEOMETRY_DEBUG environment variable to make it easier to deal with issues

Commit:
geometry_rustup: redirect cargo locate-project errors if not in a rust project

Commit:
It is okay if async_stop_worker fails because sometimes we stop before a worker is started

Commit:
geometry_rprompt: only stop async jobs if there is one and it matches geometry

Commit:
return 0 if no GEOMETRY_DEBUG

Commit:
Thank @MarioRicalde and @drager for their contributions

Commit:
Update loc

MarioRicalde committed:
README improvements.

Spellchecking, avoiding passive voice, capitalizations, punctuation, etc.

Commit:
comment on GEOMTRY_PROMPT example

MarioRicalde commented:
@justmendes I'm going to be working on #164 this week, some proposals to get the ball rolling and if someone appears that makes something better, that works too. 

I'm not sure if I'm the best person for that task, considering I'm really not top notch about everything that was implemented. I'm currently working on documenting the options, or a draft at the very least. @jedahan do you want to take on the migration function so we can release `2.0`?

Commit:
simplify function definition in contributing

Commit:
Re-render entire prompt on geometry::info, partial fix to #10

Commit:
Refactor geometry_virtualenv to read easier

justmendes commented:
I can tackle the deprecated function, no worries 👍

justmendes commented:
@jedahan do you think #226 is still worth it?

Personally I like it since it makes customization that much easier and flexible. It also allows us to reduce the number of variables and make the codebase more maintainable.

I do think if we decide to do something like that, we should first merge this and built #226 on top of it.

Commit:
Make PROMPT and RPROMPT dynamic, fixes #11

Commit:
geometry_node: use test instead of [[]]

jedahan commented:
@justmendes After merging in these changes, I think building a new version of #226 makes sense only for the git plugin. Even then, I can make examples of how to customize the git plugin simply by listing `GEOMETRY_RPROMPT=(geometry_git_conflicts geometry_git_symbol)` as an example of re-ordering without needing to change the render vars.

All other functions are so small that I feel that any customization beyond color/symbol/time variables, which we have exported, can easily be built from scratch by the user. Maybe I am not being creative enough.

MarioRicalde commented:
I agree with @jedahan - after reading the proposed structure, I can't say I'm a fan of the functional approach. I like the simplicity that the new branch offers and how clean it looks compared to the functional approach, even if it's less customizable.

Commit:
Move docker testing infrastructure to tests subdirectory

Commit:
Update license year

Commit:
Add readme logo

Commit:
Try logo before header

Commit:
Revert "Try logo before header"

This reverts commit 44629b867a653174c93f58845581b8b355dfc0b8.

Commit:
Fix screencast links

MarioRicalde committed:
options.md (#9)

* Initial options.md
* Adds missing variables to options.md thanks to @jedahan
* Adds introduction paragraph with the title.
* Removes the before segmenet in options.md
* Removes accidental double tick.
* Table formatting.
* Missing table header.
* Extra formatting.
* Adds table of contents
* Adds spacing
* Adds missing descriptions.

jedahan commented on geometry.zsh:
```diff
@@ -1,56 +1,96 @@
+: ${GEOMETRY_SEPARATOR:=" "}
+
+(($+GEOMETRY_PROMPT)) || GEOMETRY_PROMPT=(geometry_status geometry_path)
+(($+GEOMETRY_RPROMPT)) || GEOMETRY_RPROMPT=(geometry_exec_time geometry_git geometry_hg)
```
The defaults have changed significantly. I think that is okay, since it lets people know that 'things have changed', but I wonder if others feel differently and would rather it be the same out-of-box experience.

So let me know if we should change the defaults to be closer to the original


```
# move path back to left side
GEOMETRY_PROMPT=(geometry_status geometry_path)
GEOMETRY_RPROMPT=(geometry_exec_time geometry_git geometry_hg)
```

```
# add spaces around git separator again
GEOMETRY_GIT_SEPARATOR=" :: "
```

jedahan replied:
@justmendes @MarioRicalde @desyncr

jedahan reviewed (commented):
One blocker before merging

Commit:
Add branding to changelog

Commit:
Balance single space around prompt with geometry_echo

geometry_echo is used to make it easy for people to configure
if there is a space or not for either side of the prompt

Commit:
Remove options h4

Commit:
options.md: format tables

Commit:
options.md: fix spelling

Commit:
options.md: remove breaks

jedahan commented:
@MarioRicalde thats fair - let me see if theres a way to give some space, either by re-adding the breaks, or changing the header sizes or removing the `geometry_blah.zsh` being in code-formatting.

MarioRicalde commented:
What else are we pending for the release of this branch?

Cheers.

jedahan commented:
@MarioRicalde 
- [ ] [determine if we should set the same defaults as before](https://github.com/geometry-zsh/geometry/pull/238#discussion_r258964923)
- [ ] #247 (create a deprecated.zsh function) 
- [ ] #249 (write a release announcement)

MarioRicalde commented:
I can help with #1 and #3 

For #1 in your list, I think that we can keep most things as is in your branch, the only comment I have is the separators ( :: ) not having spaces before and after. Other than that it looks good.

For #3 what's the message that we want to convey? I can write a quick draft and we can go from there.

Commit:
geometry_git: Add spaces around :: for default separator

jedahan commented:
@MarioRicalde I updated #247 with my thoughts, and added the spaces around ' :: ' in the git separator

Commit:
Add semicolons to one-line functions to deal with setopt ignore_{close_}braces, fixes #239

Commit:
Embed async.zsh instead of downloading over curl, fixes #13

Commit:
Switch from async to builtin version with FS

Commit:
Do not perform prompt expansion for title setting, fixes #14

Commit:
Do not perform prompt expansion for title setting, fixes #14

MarioRicalde committed:
Adds MNML_STATUS_SYMBOL_COLOR_HASH_COLORS

Commit:
Use single parens for truthy-testing, fixes #16

Commit:
Update changelog

Commit:
Merge branch 'simple-async' into mnml

jedahan commented:
I would like to merge this branch by Wednesday, April 17th, and wait 1 week until tagging the new version. I am okay with us not having announcements ready. What do you think @desyncr @MarioRicalde @justmendes

Commit:
Fix bad pty file descriptors, fixes #241, thanks zsh-autosuggestions

Commit:
Fix bad pty file descriptors, fixes #241, thanks zsh-autosuggestions

Commit:
Merge branch 'mnml' of github.com:geometry-zsh/geometry into mnml

Commit:
Fix test Dockerfile

Commit:
Record new screencasts for git functionality and functions

Commit:
Functions last command should be echo; fixes #253 extra space in geometry_exec_time

Commit:
geometry_exec_time: each shell gets unique exec_time file

Commit:
Set screencast # color to brighter gray

Commit:
Allow different versions of containers

Commit:
Check for jobs before printing, fixes #253 for realz

Commit:
link to latest screencasts

Commit:
Rename from Geometry to mnml

Commit:
document STATUS_SYMBOL_COLOR_HASH_COLORS

Commit:
Rename geometry to mnml in changelog

This makes it a lot easier to grep and understand the history,
but I kinda feel bad about implying that this was always mnml.

I might revert this change

Commit:
move links to bottom

Commit:
Use locals wherever we can

Commit:
lowercase functions

Commit:
fix function wiki page link

frm commented:
Pinging here, based on the new updates. Have you checked [this comment](https://github.com/geometry-zsh/geometry/issues/260#issuecomment-516854407) @jedahan ?

jedahan commented:
default branch is now `mnml`, thanks for everyone's help and patience
